Common/src/main/java/mezz/jei/common/transfer/RecipeTransferUtil.java

```diff
@@ -186,23 +192,117 @@ public static RecipeTransferOperationsResult getRecipeTransferOperations(
 	) {
 		RecipeTransferOperationsResult transferOperations = new RecipeTransferOperationsResult();
 
+		// Find groups of slots for each recipe input, so each ingredient knows list of slots it can take item from
+		// and also split them between "equal" groups
+		Map<IRecipeSlotView, Map<ItemStack, ArrayList<Map.Entry<Slot, ItemStack>>>> relevantSlots = new IdentityHashMap<>();
```

Comment:
The type here is getting very long, can you refactor this into records or classes that are more descriptive?
I imagine `Map.Entry<Slot, ItemStack>` can be a record with named fields,
and something like `ArrayList<Map.Entry<Slot, ItemStack>>` could be encapsulated by a class that lets you manipulate it without exposing the underlying arraylist.

Reply:
They are `Map.Entry`ies because i passthrough them directly from map, without copying them all over place, creating heap garbage.

Reply:
The readability has been improved thanks!
I think you will find that any performance impact from the extra garbage generated here is imperceptible to players.
If we focus on optimizing as top priority, we wouldn't use Streams or Optional either, for example.

In general I think the only slowdowns I've see in JEI are from grossly bad algorithms or poorly chosen data structures, I've yet to see anything caused by extra object overhead.